src/rust/engine/src/nodes.rs

```diff
@@ -372,18 +405,47 @@ pub struct SelectDependencies {
   pub subject: Key,
   pub variants: Variants,
   pub selector: selectors::SelectDependencies,
+  pub dep_product_entries: rule_graph::Entries,
+  pub product_entries: rule_graph::Entries,
 }
 
 impl SelectDependencies {
   pub fn new(
     selector: selectors::SelectDependencies,
     subject: Key,
-    variants: Variants
+    variants: Variants,
+    edges: &rule_graph::RuleEdges
   ) -> SelectDependencies {
+    // filters entries by whether the subject type is the right subject type
+    let dep_p_entries = edges.entries_for(
+      &vec![
+      Selector::SelectDependencies(selector.clone()),
+      Selector::select(selector.clone().dep_product)
+      ]);
+    let dep_p_entries = dep_p_entries.into_iter()
+      .filter(|e| e.matches_subject_type(subject.type_id().clone()))
+      .collect::<Vec<_>>();
+
+    let p_entries = edges.entries_for(
+      &vec![
+      Selector::SelectDependencies(selector.clone()),
+      Selector::select(selector.clone().product)
+      ]);
+
+    let p_entries = p_entries.into_iter()
```

Comment:
Yeah. One of the things I realized while wiring this together is that the keys in the RuleGraph need to be better.
I can uniquely address each set of entries that fulfill a Select with a root or inner entry + a selector key which is either [Select], [OriginalSelector, InitialSelector] or, [OriginalSelector, SubjectProjection, Selector]. If the internals of the rule graph used those as keys, then this filtering wouldn't be necessary because the filter would be embedded in the key. With that, holding on to the lists wouldn't be necessary.

I was thinking of doing a fast followon change that introduces a key enum. With that, I can have the Nodes hold onto keys or key prefixes instead of Entries and they would no longer have to even know about the edge types.